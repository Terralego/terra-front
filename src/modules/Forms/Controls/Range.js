import PropTypes from 'prop-types';
import React from 'react';
import { Spinner } from '@blueprintjs/core';

import './index.scss';
import RangeNumeric from './RangeNumeric';
import RangeSlider from './RangeSlider';

export const DEFAULT_MIN = 0;
export const DEFAULT_MAX = 100;

/**
 * Range component that allows editing either by a slider or 2 numeric inputs
 */
export const Range = props => {
  const {
    label,
    min,
    max,
    threshold,
    stepSize,
    value,
    loading,
  } = props;
  const Component = (max - min) / stepSize > threshold ? RangeNumeric : RangeSlider;

  return (
    <div className="control-container control--range range">
      <p className="control-label">{label}</p>
      {loading ? <Spinner size={20} /> : (
        <Component
          {...props}
          value={(!Array.isArray(value) || value.length !== 2) ? [min, max] : value}
        />
      )}
    </div>
  );
};

Range.propTypes = {
  label: PropTypes.string,
  value: PropTypes.arrayOf(PropTypes.number),
  onChange: PropTypes.func,
  min: PropTypes.number,
  max: PropTypes.number,
  /** Number of steps for when to switch to 2 numeric inputs */
  threshold: PropTypes.number,
  stepSize: PropTypes.number,
  loading: PropTypes.bool,
};

Range.defaultProps = {
  label: '',
  value: undefined,
  onChange () {},
  min: DEFAULT_MIN,
  max: DEFAULT_MAX,
  threshold: 250,
  stepSize: 1,
  loading: false,
};

export default Range;
