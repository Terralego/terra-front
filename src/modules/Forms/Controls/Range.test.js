import { shallow, mount } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';

import Range from './Range';

jest.mock('@blueprintjs/core', () => ({
  RangeSlider: () => <p>Range</p>,
  FormGroup: ({ children }) => <div>{children}</div>,
  NumericInput: () => <p>Numeric input</p>,
  Intent: { DANGER: 'danger' },
  Spinner: () => <p>Spinner</p>,
}));

it('should render correctly loading', () => {
  const tree = renderer.create((
    <Range loading label="Pwout" />
  )).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render correctly as range slider', () => {
  const tree = renderer.create((
    <Range
      label="My range"
      value={[1, 4]}
      onChange={() => null}
      min={0}
      max={10}
    />
  )).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render correctly as two inputs', () => {
  const tree = renderer.create((
    <Range
      label="My range"
      value={[1, 4]}
      onChange={() => null}
      min={0}
      max={1000}
    />
  )).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should mount & update correctly', () => {
  const onChange = jest.fn();
  const wrapper = mount((
    <Range
      label="Change Values of my range"
      onChange={onChange}
      values={[10, 30]}
      min={0}
      max={100}
    />
  ));
  wrapper.find('RangeSlider').last().props().onRelease([20, 40]);
  expect(onChange).toHaveBeenCalled();
});

it('should render correctly with default min and max when no value', () => {
  const wrapper = shallow(<Range />);
  expect(wrapper.find('RangeSlider').last().props().value).toEqual([0, 100]);
});

it('should render correctly with default min and max if wrong value', () => {
  const wrapper = shallow(<Range value={[123]} />);
  expect(wrapper.find('RangeSlider').last().props().value).toEqual([0, 100]);
});
