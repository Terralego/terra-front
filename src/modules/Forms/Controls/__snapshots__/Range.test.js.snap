// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly as range slider 1`] = `
<div
  className="control-container control--range range"
>
  <p
    className="control-label"
  >
    My range
  </p>
  <p>
    Range
  </p>
</div>
`;

exports[`should render correctly as two inputs 1`] = `
<div
  className="control-container control--range range"
>
  <p
    className="control-label"
  >
    My range
  </p>
  <div
    className="control--range__numeric"
  >
    <div>
      <p>
        Numeric input
      </p>
    </div>
    <div>
      <p>
        Numeric input
      </p>
    </div>
  </div>
</div>
`;

exports[`should render correctly loading 1`] = `
<div
  className="control-container control--range range"
>
  <p
    className="control-label"
  >
    Pwout
  </p>
  <p>
    Spinner
  </p>
</div>
`;
